sdk/dotnet/cmd/pulumi-language-dotnet/main.go

```diff
@@ -298,55 +299,104 @@ func (host *dotnetLanguageHost) DetermineDotnetPackageDirectory(
 	return dir, nil
 }
 
-func (host *dotnetLanguageHost) DeterminePluginDependency(
-	ctx context.Context, packageDir, packageName, packageVersion string) (*pulumirpc.PluginDependency, error) {
+type versionFile struct {
+	name    string
+	version string
+}
+
+func newVersionFile(b []byte, packageName string) *versionFile {
+	var name string
+	version := strings.TrimSpace(string(b))
+	parts := strings.SplitN(version, "\n", 2)
+	if len(parts) == 2 {
+		// version.txt may contain two lines, in which case it's "plugin name\nversion"
+		name = strings.TrimSpace(parts[0])
+		version = strings.TrimSpace(parts[1])
+	}
+
+	if !strings.HasPrefix(version, "v") {
+		// Version file has stripped off the "v" that we need. So add it back here.
+		version = fmt.Sprintf("v%v", version)
+	}
+
+	return &versionFile{
+		name:    name,
+		version: version,
+	}
+}
+
+func DeterminePluginDependency(packageDir, packageName, packageVersion string) (*pulumirpc.PluginDependency, error) {
 
 	logging.V(5).Infof("GetRequiredPlugins: Determining plugin dependency: %v, %v, %v",
 		packageDir, packageName, packageVersion)
 
 	// Check for a `~/.nuget/packages/package_name/package_version/content/version.txt` file.
 
-	versionFilePath := path.Join(packageDir, strings.ToLower(packageName), packageVersion, "content", "version.txt")
+	artifactPath := filepath.Join(packageDir, strings.ToLower(packageName), packageVersion, "content")
+	pulumiPluginFilePath := filepath.Join(artifactPath, "pulumiplugin.json")
+	versionFilePath := filepath.Join(artifactPath, "version.txt")
+	logging.V(5).Infof("GetRequiredPlugins: plugin file path: %v", versionFilePath)
 	logging.V(5).Infof("GetRequiredPlugins: version file path: %v", versionFilePath)
 
-	if _, err := os.Stat(versionFilePath); err != nil {
-		if os.IsNotExist(err) {
-			// Pulumi package doesn't contain a version.txt file.  This is not a resource-plugin.
-			// just ignore it.
-			logging.V(5).Infof("GetRequiredPlugins: No such file")
-			return nil, nil
-		}
-
-		// some other error.  report it as it means we can't read this important file.
-		logging.V(5).Infof("GetRequiredPlugins: err: %v", err)
+	pulumiPlugin, err := plugin.LoadPulumiPluginJSON(pulumiPluginFilePath)
+	if os.IsNotExist(err) {
+		pulumiPlugin = nil
```

Comment:
Is this necessary? Won't `pulumiPlugin` already be `nil` when `err != nil` and therefore `os.IsNotExists(err)` is `true`?

Reply:
We need the check, but not the assignment. I can fold it down to this
```go
if err != nil && !os.IsNotExist(err) {
		return nil, err
	}
```